### PR TITLE
 #90: refactor tests

### PR DESCRIPTION
**Pull request by silkentrance**
This refactors existing tests into separate test modules so one can run tests individually.
In addition this adds the eat-your-own-dog-food feel to the test suite.

## Changes in detail

- add nyc to the mix
- improve travis configuration to test against not just the latest node release
- check coverage during travis build (we have some tests missing there)
- improve error message when inheriting from another ``@suite`` decorated class
- add work around in test/it/util.ts#cleanup() when running tests with nyc (see the comment in the file)
- refactor all tests into separate entities so that they can be run separately
- abstract test bases must not cleanup the stdout/stderr before passing it to assertOutput() (redundant)
- add itrun script so one can run individual test suites during development, e.g. ``npm run itrun test/it/DependencyInjectionSuite.ts``, without having to run all of the existing other test suites including the slow package test suites
- improve travis config to not build both the branch (named gh-.*) and the PR that belongs to it, just build the PR.

## Good News

Overall coverage is very good:

```
Statements: 91.78% ( 346/377 )
Branches: 80.95% ( 170/210 )
Functions: 97% ( 97/100 )
Lines: 91.83% ( 337/367 )
```
but we are missing some tests in order to increase overall coverage.

## Bad News

- NodeJS 6 builds are failing in the PackageIT tests, too, it seems that the parameterisation is wrong for packaging the mocha-typescript package and testing against test/it/fixtures/packages/* on whether it can be used. (something about npm not being able to install from a tgz.)
- NodeJS 4 builds are constantly failing - removed that target from travis conf

## TODO

- [x] Dependency Injection test suite
- [x] cleanup code
- [ ] documentation
- [x] clean up / extend npm scripts

### CODE REVIEW

**Comment by silkentrance** [comment]
@pana-cc can we also have the tests run on appveyor? Most of the users use the Windows OS and need some confidence that this is also running on their platform.

**Comment by silkentrance** [comment]
@pana-cc please review as I am done with the implementation part, as for documentation, it is up to you to decide whether we include documentation on how to contribute/develop.

File: .travis.yml [hunk 1]
```diff
@@ -1,3 +1,24 @@
 language: node_js
 node_js:
-  - "node"
\ No newline at end of file
+  - "node"
+  - "10"
+  - "9"
```
Comment (pana-cc) [inline .travis.yml]: 👍

File: index.ts [hunk 2]
```diff
@@ -216,33 +217,37 @@ function suiteClassCallback(target: SuiteCtor, context: TestFunctions) {
...
 
-                    applyTestFunc(testFunc, testName, method, [], method.length === 0);
+                    applyTestFunc(testFunc, testName, method, []);
                 }
             }
         }
 
+        function isAsync(method: Function): boolean {
+
+            return RE_ASYNC.test(method.toString());
```
Comment (pana-cc) [inline index.ts]: 👎 Can't this still use length? What does mocha use to check if a function is async?

File: test/it/fixtures/test.suite.ts [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+// //   tear down server.
+//
+// @suite @timeout(10) class OverallSlow {
+//     @test public "first fast"(done) {
+//         setTimeout(done, 1);
+//     }
+//     @test public "second slow"(done) {
+//         setTimeout(done, 20);
+//     }
+//     @test public "third fast"(done) {
```
Comment (pana-cc) [inline test/it/fixtures/test.suite.ts]: I wonder if I let all this commented tests slip in a PR

File: index.ts (hunk 2 above)
Comment (silkentrance) [reply index.ts]: ups wonder why this is still in there...
mocha just uses the length.
but for our purposes, especially with parameter dependency injection and async tests, I think that we will need this and some convention on how to name the `done` parameter.
I will remove this for now.

File: test/it/fixtures/test.suite.ts (hunk 3 above)
Comment (silkentrance) [reply test/it/fixtures/test.suite.ts]: reactivate?

File: index.ts (hunk 2 above)
Comment (silkentrance) [reply index.ts]: the problem is with @params, need to check whether the decorator is available...

Comment (silkentrance) [reply index.ts]: fixed

File: test/it/fixtures/test.suite.ts (hunk 3 above)
Comment (silkentrance) [reply test/it/fixtures/test.suite.ts]: reactivated - guess what, the coverage just went up 👍

File: index.ts (hunk 2 above)
Comment (silkentrance) [reply index.ts]: 🦅 eye... great!

File: test/it/fixtures/test.suite.ts (hunk 3 above)
Comment (silkentrance) [reply test/it/fixtures/test.suite.ts]: 🦅 eye... great!

File: .travis.yml (hunk 1 above)
Comment (silkentrance) [reply .travis.yml]: v5 + v6 continue to fail, v5 started to fail once the below fixed had been applied incl. reactivating the priorly disabled test suites.
I will remove both versions incl. also 7 from the travis build.

**Comment by pana-cc** [comment]
BTW, Congratulations, you have the honor to merge PR number 100. 🎉

**Comment by pana-cc** [comment]
Downloads doesn't look bad
https://npm-stat.com/charts.html?package=mocha-typescript

**Comment by silkentrance** [comment]
@pana-cc well the weekly download drop well below the 5k mark. I'd like them to be around >20k. Perhaps making some advertisement?

**Comment by silkentrance** [comment]
@pana-cc since this might affect the existing release, can you please publish a new release. I am still in the process of figuring out on how to do it myself. And before I break anything... especially with npm limiting releases and not allowing to overwrite existing releases.

BTW can you point me towards a working documentation on how to publish new releases to npm as a contributor?

**Comment by pana-cc** [comment]
I've added the steps I take to publish the package: https://github.com/pana-cc/mocha-typescript/pull/105

**Pull request merged**